### PR TITLE
Use upstream rocksdb

### PR DESCRIPTION
**Pull request by dvdplm**
…by way of https://github.com/paritytech/parity-common/pull/257 by @ordian.

### CODE REVIEW

**Comment by grbIzl** [comment]
(Sorry for the stupid question) Do we have any metric or tests for it? I mean, do we understand, how will this affect client's performance? @ordian

**Comment by dvdplm** [comment]
> Do we have any metric or tests for it?

Sort of. Both me and @ordian have been running block import tests over the last few days. Here are mine from last night, importing 5k blocks from approx mid-chain:

Master:
```
Benchmark #1: ./parities/parity-master --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     320.989 s ± 24.305 s    [User: 341.816 s, System: 41.915 s]
  Range (min … max):   302.979 s … 362.882 s    5 runs   
```

Upstream rocksdb:
```
Benchmark #1: ./parities/parity-ao --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
   Time (mean ± σ):     283.643 s ± 10.213 s    [User: 282.681 s, System: 43.740 s]
   Range (min … max):   271.713 s … 296.393 s    5 runs
```

I think @ordian's results are similar.

**Comment by ordian** [comment]
@grbIzl I'm seeing 20% improvement on importing blocks `parity import` when switching to this branch (tested on 100 and 5k recent blocks from mainnet at ~8.9m).

**Comment by dvdplm** [comment]
More benchmarks, now also comparing master from today (Nov 20, 2019) after the new trie landed (referred to below as "fast trie", but it's actually a little slower, cc @cheme @jimpo):

**Upstream rocksdb**
```
Benchmark #1: ./parities/parity-ao --mode=offline --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     284.806 s ±  7.885 s    [User: 283.799 s, System: 42.353 s]
  Range (min … max):   277.414 s … 296.277 s    5 runs
```

**Upstream rocksdb + fast trie**
```
Benchmark #1: ./parities/parity-ao-fast-trie --mode=offline --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     292.625 s ± 12.209 s    [User: 290.013 s, System: 42.446 s]
  Range (min … max):   282.579 s … 312.988 s    5 runs
```

**Forked rocksdb (master, Nov 20)**
```
Benchmark #1: ./parities/parity-master --mode=offline --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     321.283 s ±  8.235 s    [User: 340.501 s, System: 41.963 s]
  Range (min … max):   312.589 s … 333.771 s    5 runs
```

**Forked rocksdb + fast trie (master, Nov 21)**
```
Benchmark #1: ./parities/parity-master-fast-trie --mode=offline --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     323.682 s ±  7.001 s    [User: 341.980 s, System: 40.867 s]
  Range (min … max):   317.452 s … 334.709 s    5 runs
```

I also ran the same set of benchmarks with more verifier threads (8), but it has little impact.

**Upstream rocksdb + 8 verif threads**
```
Benchmark #1: ./parities/parity-ao --mode=offline --scale-verifiers --num-verifiers=8 --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     277.217 s ±  8.206 s    [User: 266.063 s, System: 41.339 s]
  Range (min … max):   268.896 s … 288.652 s    5 runs
```

**Upstream rocksdb + fast trie + 8 verif threads**
```
Benchmark #1: ./parities/parity-ao-fast-trie --mode=offline --scale-verifiers --num-verifiers=8 --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     288.779 s ±  8.539 s    [User: 274.010 s, System: 41.654 s]
  Range (min … max):   279.360 s … 301.326 s    5 runs
```

**Forked rocksdb + 10 verif threads (master, Nov 20)**
```
Benchmark #1: ./parities/parity-master --mode=offline --scale-verifiers --num-verifiers=8 --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     311.093 s ±  5.957 s    [User: 323.145 s, System: 40.403 s]
  Range (min … max):   301.396 s … 316.075 s    5 runs
```

**Forked rocksdb + fast trie + 8 verif threads (master, Nov 21)**
```
Benchmark #1: ./parities/parity-master-fast-trie --mode=offline --chain=eth --db-path=./current/bench-mainnet --pruning-history=12000 import ./blocks/blocks-5000.bin
  Time (mean ± σ):     322.392 s ±  7.124 s    [User: 341.089 s, System: 40.680 s]
  Range (min … max):   314.879 s … 333.632 s    5 runs
```

**Comment by cheme** [comment]
That's very interesting, trie db crate improvement are mainly on triedb iterator,
it is possible that it does not matter a lot for ethereum.
I suspect the loss come from decoding where we first build a nodeplan and then call 'build' function to go back to 'Node'. Before the pr 'Node' was directly produce from 'decode'. IIrc perf degrades on trie iter the same way at first but @jimpo fix that by avoiding 'build' calls it gets better. Probably it is the same here: we should avoid calling 'build' in triedbmut by having 'Node' replaced by some struct like '(data_slice, NodePlan)' but it is just a wild guess.

It could be rather simple to test: just put back the old 'decode' function in node_codec and see if it fix the perf (decode still exist in the trait but with a default implementation that call the 'build' function).

**Comment by grbIzl** [comment]
@ordian @dvdplm thanks for info! Looks promising. May be worth mentioning in PR's description also.

**Comment by dvdplm** [comment]
> It could be rather simple to test: just put back the old 'decode' function in node_codec and see if it fix the perf (decode still exist in the trait but with a default implementation that call the 'build' function).

I did this and but my micro-benchmark did not show any meaningful regression, not sure if my benchmark is bad perhaps. :/

**Comment by cheme** [comment]
> I did this and but my micro-benchmark did not show any meaningful regression, not sure if my benchmark is bad perhaps. :/

Thanks for checking :+1: , should be from somewhere else then :\ .

**Comment by General-Beck** [comment]
I think it will be necessary to test the build on the architectures aarch64, armv7, i386. And also on Windows and Mac platforms.
`parity-clib` library for android needs a separate test on the android device.

**Comment by General-Beck** [comment]
build test's https://gitlab.parity.io/General-Beck/parity-ethereum/pipelines/70139
CI updated, on Linux hosts installed clang-8, on windows clang-9

**Comment by dvdplm** [comment]
> Linux hosts installed clang-8, on windows clang-9

Why not the same version on both? i.e. clang-9?

**Comment by General-Beck** [comment]
> Why not the same version on both? i.e. clang-9?
During the discussion phase, I and @TriplEight  do not mind using stable clang-9. We only need confirmation from the developers. and so that it does not interfere with other projects

**Comment by dvdplm** [comment]
> confirmation from the developers. and so that it does not interfere with other projects

Hmm, how do you mean, "developers"? And I'm not sure which other projects share the same CI setup, can you elaborate?

**Comment by General-Beck** [comment]
@dvdplm about our projects that use the same images in СI. then go to `clang-9`

**Comment by TriplEight** [comment]
Let me clarify. We've decided to move **all** our CI environments to `clang-9`.

**Comment by dvdplm** [comment]
> Let me clarify. We've decided to move **all** our CI environments to `clang-9`.

Please let us know when the work is completed, thanks!

**Comment by General-Beck** [comment]
build passed

File: parity/db/rocksdb/helpers.rs [hunk 1]
```diff
@@ -27,10 +27,38 @@ pub fn compaction_profile(profile: &DatabaseCompactionProfile, db_path: &Path) -
...
 }
 
+pub fn memory_per_column(total: Option<usize>) -> HashMap<Option<u32>, usize> {
+	let mut memory_per_column = HashMap::new();
+	if let Some(budget) = total {
+		// spend 90% of the memory budget on the state column, but at least 256 MiB
+		memory_per_column.insert(ethcore_db::COL_STATE, std::cmp::max(budget * 9 / 10, 256));
+		let num_columns = ethcore_db::NUM_COLUMNS.expect("NUM_COLUMNS is Some; qed");
+		// spread the remaining 10% evenly across columns
+		let rest_budget = budget / 10 / (num_columns as usize - 1);
```
Comment (niklasad1) [inline parity/db/rocksdb/helpers.rs]: why `num_columns as usize - 1` here?

Comment (ordian) [reply parity/db/rocksdb/helpers.rs]: -1 for the state column

File: parity/db/rocksdb/mod.rs [hunk 2]
```diff
@@ -86,16 +86,16 @@ pub fn restoration_db_handler(client_path: &Path, client_config: &ClientConfig)
...
+/// Open a new light client DB.
+pub fn open_db_light(
 	client_path: &str,
 	cache_config: &CacheConfig,
 	compaction: &DatabaseCompactionProfile
 ) -> io::Result<Arc<dyn BlockChainDB>> {
 	let path = Path::new(client_path);
 
 	let db_config = DatabaseConfig {
-		memory_budget: Some(cache_config.blockchain() as usize * 1024 * 1024),
```
Comment (niklasad1) [inline parity/db/rocksdb/mod.rs]: hmm, was this is a bug? `CacheConfig` stores all caches as MB IIRC

File: parity/db/rocksdb/helpers.rs (hunk 1 above)
Comment (ordian) [reply parity/db/rocksdb/helpers.rs]: ah, it should probably be -1 for the light db as well

File: parity/db/rocksdb/helpers.rs [hunk 3]
```diff
@@ -27,10 +27,38 @@ pub fn compaction_profile(profile: &DatabaseCompactionProfile, db_path: &Path) -
...
+		}
+	}
+	memory_per_column
+}
+
+pub fn memory_per_column_light(total: usize) -> HashMap<Option<u32>, usize> {
+	let mut memory_per_column = HashMap::new();
+	let num_columns = ethcore_db::NUM_COLUMNS.expect("NUM_COLUMNS is Some; qed");
+	// spread the memory budget evenly across columns
+	let per_column = total / (num_columns as usize);
```
Comment (ordian) [inline parity/db/rocksdb/helpers.rs]:
```suggestion
	let per_column = total / (num_columns as usize - 1);
```

File: parity/db/rocksdb/helpers.rs [hunk 4]
```diff
@@ -27,10 +27,38 @@ pub fn compaction_profile(profile: &DatabaseCompactionProfile, db_path: &Path) -
 	}
 }
 
+pub fn memory_per_column(total: Option<usize>) -> HashMap<Option<u32>, usize> {
```
Comment (niklasad1) [inline parity/db/rocksdb/helpers.rs]: total is expected to be in `MB`? We should document it.

File: parity/db/rocksdb/helpers.rs [hunk 5]
```diff
@@ -27,10 +27,38 @@ pub fn compaction_profile(profile: &DatabaseCompactionProfile, db_path: &Path) -
...
+		let rest_budget = budget / 10 / (num_columns as usize - 1);
+		for i in 1..num_columns {
+			// but at least 16 MiB for each column
+			memory_per_column.insert(Some(i), std::cmp::max(rest_budget, 16));
+		}
+	}
+	memory_per_column
+}
+
+pub fn memory_per_column_light(total: usize) -> HashMap<Option<u32>, usize> {
```
Comment (niklasad1) [inline parity/db/rocksdb/helpers.rs]: total is expected to be in MB? We should document it.

File: parity/db/rocksdb/mod.rs (hunk 2 above)
Comment (ordian) [reply parity/db/rocksdb/mod.rs]: yes, it seems light client didn't respect the memory budget

**Comment by ordian** [comment]
> Do we have some benchmarks?

I talked to the devops team to set up running the import benches on a daily basis or on-demand at least to track perf regression, will let you know once it's done.

**Pull request merged**